azurelinuxagent/ga/exthandlers.py

```diff
@@ -927,19 +913,25 @@ def handle_disable(self, ext_handler_i, extension=None):
         if handler_state == ExtHandlerState.Enabled:
             ext_handler_i.disable(extension)
 
-    def handle_uninstall(self, ext_handler_i, extension):
+    def handle_uninstall(self, ext_handler_i, extension, policy_engine):
         """
         To Uninstall the handler, first ensure all extensions are disabled
             1- Disable all enabled extensions first if Handler is Enabled and then Disable the handler
                 (disabled extensions wont have any extensions dependent on them so we can just go
                 ahead and remove all of them at once if HandlerState==Uninstall.
                 CRP will only set the HandlerState to Uninstall if all its extensions are set to be disabled)
             2- Finally uninstall the handler
+
+        'policy_engine' should be an ExtensionPolicyEngine object, which is invoked to determine if the extension is allowed by policy.
+        Note: if the extension was never installed, uninstall is allowed even for disallowed extensions because no extension code will be executed.
         """
         handler_state = ext_handler_i.get_handler_state()
         ext_handler_i.logger.info("[Uninstall] current handler state is: {0}", handler_state.lower())
         if handler_state != ExtHandlerState.NotInstalled:
             if handler_state == ExtHandlerState.Enabled:
+                # Check policy and raise an error if the extension is disallowed (e.g., not in allowlist, signature not previously validated when required).
+                policy_engine.enforce_policy_for_ext_handler(ext_handler_i)
```

Comment:
We only enforce policy here but if we don't enter this block then we attempt to uninstall (invokes ext code) on line 947 without checking the policy first